Parallelize CheckInputs() in AcceptToMemoryPool()

sdaftuar opened the pull request:
This PR takes advantage of our script check threads (used for block validation) in order to speed up transaction validation when adding to the mempool.

The motivation is to minimize the amount of time we might spend validating a single transaction.  I think there are several approaches we could take to mitigate this concern, but this approach seemed straightforward to implement, and would be easy to revert if we later opted for a different solution (such as parallelizing our message handler or dropping transactions into some sort of queue for background processing).

This PR does introduce some behavior changes:
- we would no longer ban peers for transactions with invalid scripts (however, the status quo behavior is spotty and inconsistent in how it bans anyway, since the mandatory consensus flags set is years old and we stop at the first script execution error)
- we could relay transactions from a whitelisted peer that would cause us to be banned by an older peer (but `-whitelistforcerelay` now defaults to off, mitigating this concern)

sdaftuar committed:
Change CheckInputs() to only validate scripts once

Previously CheckInputs() could invoke a script twice to determine if a script
failure was due to standardness rules. However, it only performed this in
single-threaded validation mode, and it directly accessed policy variables to
make this determination (even though the caller was passing in the desired
script flags to use).

The policy vs consensus script flags comparison seems out of touch with current
network deployments; the only MANDATORY script flag is P2SH, even though many
other consensus changes have been deployed since. However, adding further flags
to the MANDATORY set risks making the p2p behavior overly restrictive by
disconnecting unupgraded peers unnecessarily.

Moreover, AcceptToMemoryPoolWorker() already invokes CheckInputs multiple
times, with different script flags, to detect certain kinds of segwit
validation errors. Given that the caller is responsible for passing in the
correct validation flags, it feels like a layer violation for CheckInputs() to
be re-running checks with different script flags by itself.

This patch moves all responsibility for determining which script flags are
mandatory to the caller, so CheckInputs() will now always return CONSENSUS for
any script failures. The caller can rewrite the error code itself, eg if
policy-flags were included.

This patch also changes the ban behavior for peers sending transactions with
invalid scripts; we will no longer disconnect such peers. It was already
trivial for an adversary to mutate a transaction to violate a policy rule,
previously, in order to relay a transaction and avoid a ban, so we were not
relying on the ban behavior previously.

As this results in changes to the reject reasons and codes we can return, many
of the tests have been updated as well.

sdaftuar committed:
Parallelize script validation in AcceptToMemoryPool()

When receiving transactions over the network, use script check threads (if
available) to reduce worst-case message handling time.

Continue to use single-threaded validation for rpc invocations (which results
in more descriptive error messages, on failure), where DoS is not a significant
concern.

gmaxwell commented:
> we could relay transactions from a whitelisted peer that would cause us to be banned by an older peer

This seems bad.  The original motivation for making whitelist relay is long long long since dead (*).  Can we just eliminate this behaviour, as it makes whitelisting less useful and isn't useful itself?

If there is still any need for forced-relay transactions it could be accomplished by introducing a new P2P message that only whitelisted peers are allowed to use, that carries with it a TX whos relay is forced.

(*) my recollection is that the original reason that behaviour was added was because people using armory would start armory and their bitcoin node at the same time, and their armory wallet would blast txn to send at bitcoind which would end up in its mempool but not relayed on the network because it had no actual peers up. ... then later attempts by armory to rebroadcast were ineffectual because the txn were already in the node's mempool.   However, because it causes all your peers to be flooded by txn relayed by your whitelisted neighbours it makes whitelisting useless outside of that armory usecase, unless you add the second option to force relay off.

gmaxwell commented:
On the overall change, sounds interesting but without benchmarks I dunno how exciting it is.  I guess I'm kinda surprised that there is a speedup at all considering that IBD doesn't seem to scale past 15/16 cores... which I'd assumed was due to synchronization overheads in the script validation thread dispatch.

sdaftuar commented:
I would love to get rid of the whitelist relay behavior.  Any opinions about whether I should open a new PR for that, or just do it here?

My initial benchmark is on a very large transaction that has 640 P2PKH inputs, 2 outputs, and is policy-valid.  I compared this PR with master, and I tried this on two computers, a Power9 workstation (which doesn't have the SHA optimizations) and my laptop (a recent macbook pro).

With master (verification threads are unused in mempool acceptance):

| Computer | Verification Threads | average scriptcheck time | average total atmp time|
|-----------|---------------------|------------------|----------------|
|power9|16|283.0ms|285.6ms|
|macbook pro|8|176.7ms|178.5ms|

The "scriptcheck time" is a timing I added from just before the first call to CheckInputs(), to just after the second call.  (I compare with the total ATMP time to sanity check that this is a meaningful optimization.)

With this PR:

| Computer | Verification Threads | average scriptcheck time | average total atmp time|
|-----------|---------------------|------------------|----------------|
|power9|16|35.7ms|38.2ms|
|macbook pro|8|40.7ms|42.6ms|

I'm actually somewhat more interested in benchmarking transactions that use maximal CPU resources but don't get into the mempool, since such transactions are costless for someone to generate (and because they're slower for us to process, at least in this PR as it stands) -- I will try to post those numbers later.  Also, if you have any suggestions for transactions that might be more costly to validate than what I've done so far, please let me know.

JeremyRubin reviewed (commented):
concept ack.

I think I'd rather see the parallelization determination be abstracted into the CheckQueue itself; although I think it's difficult to envision how that could be done neatly without incurring the overhead of constructing the check jobs. Maybe what would be best is to make it so that the overhead of using the queue is sufficiently low such that we don't need to worry about only using it for a certain number of inputs. IE, someday if the queue is reworked such that this overhead goes away, it would be weird that code gating the use of the queue dependent on job parallelization  exists outside the queue.

That said, can you provide a benchmark detailing how bad it is to just always use the queue as-is?

gmaxwell commented:
@sdaftuar on forced relay, I think a PR that defaulted it to off could just be open separately and make it in faster than this work... and this would could continue assuming that issue is resolved elsewhere.

sdaftuar commented on src/validation.cpp:
```diff
@@ -1679,6 +1680,30 @@ void ThreadScriptCheck() {
+    bool parallel_verify = false;
+    if (legacy_inputs >= 15 && nScriptCheckThreads) {
+        parallel_verify = true;
+    }
```
I think you're right that it's not worth the complexity of gating the logic on the number of legacy inputs.  I did some microbenchmarking on a couple of my computers to see how long it takes to validate small transactions at various thresholds, and the effect seems very small.

sdaftuar commented:
Updated to always use parallel validation for mempool transactions if scriptcheck threads are available.

One remaining open question in my mind is around the new behavior where a peer sending an invalid-according-to-consensus rules transaction would no longer be disconnected.  The motivation for this  is to mitigate adversarial behavior: in order to determine whether a transaction is invalid according to consensus rules (rather than policy) we end up validating a transaction's scripts twice, so an adversary trying a CPU DoS could construct transactions that only fail policy checks, but otherwise tie up our CPU for twice as long.

However, in a non-adversarial case, where our peer is merely broken (or intended to operate on a different network) but non-malicious, this means that we will stay connected to the peer longer than we should.  This is also not ideal.  I have two ideas for addressing this, but I welcome other suggestions: 

(1) probabilistically evaluate failing transactions with consensus flags, and using the result to decide whether to disconnect a peer -- say we had a 1% chance of validating a failed transaction a second time, so that if a peer is giving us garbage we can figure it out relatively quickly while still limiting CPU usage for a bad transaction in the general case.

(2) @TheBlueMatt suggested I refactor the checkqueue logic to pass back the reason for a script failure (say, the first script that fails when doing parallel validation), and then we use that to determine whether to disconnect a peer based on a single validation of scripts.  Assuming that there are no conceptual issues with this, then this should be optimal behavior for us.  This also seems a little invasive and IMO the gain here is not obviously worth the effort, but Matt suggested we might want this the next time a consensus change is proposed anyway.

Any thoughts on whether the behavior change proposed here is ok, or if I should pursue an alternate approach, such as one of the two ideas above?

JeremyRubin commented:
The check queue approach is IMO a bad idea because we may fail for different reasons on different executions.

If you want a reason, I would re-run validation in single core and reproduce the first failure.

Otherwise maybe you can produce all failure tuples (index, failure) and then take the sorted first one?

JeremyRubin commented:
You can even track the min txn that any worker has attempted and then just restart from there -- that might save you some work against blocks which have their last txn fail.

sdaftuar commented:
@JeremyRubin For a transaction containing multiple failing script inputs, I don't think it should matter whether the failure we associate with it is deterministic, since trying to find the reason for validation failure is only something we're trying to do for a non-adversarial case anyway. Do you have a scenario in mind where it would be an issue?

sdaftuar committed:
[refactor] Eliminate default argument to ATMP

Default arguments make review easier when adding a single new argument to a
widely called function, but then are harder to keep track of (and ensure
correctness of callers) when adding new arguments in the future.

sdaftuar commented:
I added a commit (to be squashed) which uses single-threaded validation for transactions submitted over RPC, since we're not worried about DoS there and users can get more detailed error information for invalid transactions when we scripts are evaluated serially.  @jnewbery -- thanks for the suggestion.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -895,6 +895,20 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
         // This is done last to help prevent CPU exhaustion denial-of-service attacks.
         PrecomputedTransactionData txdata(tx);
         if (!CheckInputs(tx, state, view, true, scriptVerifyFlags, true, false, txdata)) {
+            // Assume that this is due to a policy violation, rather than
```
Hmm, one thing this potentially protects us from is inbound peers on fork chains where the chain doesn't (yet) have any blocks to relay us (so that we ban them for invalid headers). We may want to broaden the ConsiderEviction logic to apply some similar (but relaxed - don't want to disconnect peers which are in IBD and making progress) rules to inbound peers. No need to do it in this PR, though - I think our inbound peer eviction logic could use some tuning independently and no huge risk even in the case of more fork chains given the outbound peer eviction logic robustness.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -895,6 +895,20 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
+            // Anyway re-checking the scripts with a different set of flags
+            // wastes CPU as well (particularly if an attacker is constructing
+            // transactions designed to waste our CPU).
+            state.Invalid(ValidationInvalidReason::TX_NOT_STANDARD, false, REJECT_NONSTANDARD, state.GetRejectReason(), state.GetDebugMessage());
```
nit: maybe include in the comment why we have to call this function anyway (that CheckInputs always returns CONSENSUS so we have to overwrite it).

sdaftuar replied:
Done

TheBlueMatt reviewed (approved):
utACK b689d4de467cf4031a2473afc08d2a0fe1b6c71b. Some comments but none all that important, really think this is a more important change than it gets credit for.

sdaftuar commented:
Thanks for the review @TheBlueMatt.  I'll hold off on the comment nit unless I end up updating this PR for something else.

jnewbery commented on test/functional/feature_segwit.py:
```diff
@@ -154,8 +154,8 @@ def run_test(self):
         self.log.info("Verify unsigned p2sh witness txs without a redeem script are invalid")
-        self.fail_accept(self.nodes[2], "mandatory-script-verify-flag", p2sh_ids[NODE_2][WIT_V0][1], False)
-        self.fail_accept(self.nodes[2], "mandatory-script-verify-flag", p2sh_ids[NODE_2][WIT_V1][1], False)
+        self.fail_accept(self.nodes[2], "script-verify-flag", p2sh_ids[NODE_2][WIT_V0][1], False)
```
The failure string can be "script-verify-flag-failed" here. Even better, with the specific error string "script-verify-flag-failed (Operation not valid with the current stack size)"

sdaftuar replied:
Leaving this one alone -- someone interested in improving this test can take it on in a separate PR.

jnewbery reviewed (commented):
A few small comments on the test changes.

A few comments on the git commits:
- I recommend removing the word 'Refactor' from the commit log for _Refactor CheckInputs() to only validate scripts once_ (since it's not a refactor - it's a change in behaviour)
- Are you able to update the commit log for _[refactor] Eliminate default argument to ATMP_ to explain why you're making this change?
- Can you squash the squashme commit?

sdaftuar commented on test/functional/feature_dersig.py:
```diff
@@ -101,8 +101,8 @@ def run_test(self):
-            [{'txid': spendtx.hash, 'allowed': False, 'reject-reason': '64: non-mandatory-script-verify-flag (Non-canonical DER signature)'}],
-            self.nodes[0].testmempoolaccept(rawtxs=[spendtx.serialize().hex()], maxfeerate=0)
+            [{'txid': spendtx.hash, 'allowed': False, 'reject-reason': '64: script-verify-flag-failed (Non-canonical DER signature)'}],
+            self.nodes[0].testmempoolaccept(rawtxs=[spendtx.serialize().hex()], allowhighfees=0)
```
Fixed

sdaftuar commented on test/functional/feature_cltv.py:
```diff
@@ -112,16 +112,16 @@ def run_test(self):
-            [{'txid': spendtx.hash, 'allowed': False, 'reject-reason': '64: non-mandatory-script-verify-flag (Negative locktime)'}],
-            self.nodes[0].testmempoolaccept(rawtxs=[spendtx.serialize().hex()], maxfeerate=0)
+            [{'txid': spendtx.hash, 'allowed': False, 'reject-reason': '64: script-verify-flag-failed (Negative locktime)'}],
+            self.nodes[0].testmempoolaccept(rawtxs=[spendtx.serialize().hex()], allowhighfees=0)
```
Fixed

sdaftuar commented on test/functional/p2p_segwit.py:
```diff
@@ -184,7 +184,7 @@ def set_test_params(self):
         self.num_nodes = 3
         # This test tests SegWit both pre and post-activation, so use the normal BIP9 activation.
-        self.extra_args = [["-whitelist=127.0.0.1", "-vbparams=segwit:0:999999999999"], ["-whitelist=127.0.0.1", "-acceptnonstdtxn=0", "-vbparams=segwit:0:999999999999"], ["-whitelist=127.0.0.1", "-vbparams=segwit:0:0"]]
+        self.extra_args = [["-whitelist=127.0.0.1", "-vbparams=segwit:0:999999999999", "-whitelistforcerelay=0"], ["-whitelist=127.0.0.1", "-acceptnonstdtxn=0", "-vbparams=segwit:0:999999999999", "-whitelistforcerelay=0"], ["-whitelist=127.0.0.1", "-vbparams=segwit:0:0", "-whitelistforcerelay=0"]]
```
Removed

sdaftuar commented:
Addressed all but one of @jnewbery's comments, and included the comment change that @TheBlueMatt previously suggested.

Prior version of this PR can be compared here: [15169.1](https://github.com/sdaftuar/bitcoin/commits/15169.1)

jnewbery reviewed (commented):
A few comments inline. One more:

- can you remove `MANDATORY_SCRIPT_VERIFY_FLAGS` and its comment from policy.h (and replace it in the definition of  `STANDARD_SCRIPT_VERIFY_FLAGS` with `SCRIPT_VERIFY_P2SH`) now that it no longer has any meaning?

sdaftuar committed:
Remove unused MANDATORY_SCRIPT_VERIFY_FLAGS

This variable is no longer used, so remove it.

Previously, this was used by validation logic to determine whether a
transaction with invalid scripts was violating a policy rule or a consensus
rule -- the idea being that we might ban a peer that relays an
invalid-according-to-consensus rule, but not an invalid-according-to-policy
one.

However, consensus rule changes are designed so that nodes need not all upgrade
at once. Consequently, banning a peer for relaying a transaction that is newly
invalid as a result of a recent soft-fork would risk partitioning old nodes
from the network needlessly. Because of this type of concern, the only consensus
flag every added to the MANDATORY set was P2SH.

sdaftuar commented:
@jnewbery Good catch on the caching bug.  I've pushed up a quick fix for now, but probably I should do something better here when I have a chance to revisit.

fjahr commented on src/rpc/rawtransaction.cpp:
```diff
@@ -925,7 +925,7 @@ static UniValue testmempoolaccept(const JSONRPCRequest& request)
         LOCK(cs_main);
         test_accept_res = AcceptToMemoryPool(mempool, state, std::move(tx), &missing_inputs,
-            nullptr /* plTxnReplaced */, false /* bypass_limits */, max_raw_tx_fee, /* test_accept */ true);
+            nullptr /* plTxnReplaced */, false /* bypass_limits */, max_raw_tx_fee, /* test_accept */ true, /* single_threaded_validation */ true);
```
Style-Nit: Would be a bit nicer to have consistency in the ordering of argument and comment.

MarcoFalke replied:
The arguments might be wrapped into a struct before they are passed at some point in the future, so I guess this style inconsistency will be solved at that point

fjahr commented on src/validation.cpp:
```diff
@@ -1672,6 +1675,23 @@ void ThreadScriptCheck(int worker_num) {
     scriptcheckqueue.Thread();
 }
 
+bool RunCheckInputsMaybeParallel(const CTransaction& tx, CValidationState &state, const CCoinsViewCache &inputs, unsigned int flags, bool cacheSigStore, bool cacheFullScriptStore, PrecomputedTransactionData& txdata, bool single_threaded_validation) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```
Question on code structure: I would like to understand the reasoning for wrapping CheckInputs() in RunCheckInputsMaybeParalell(). Rather than that I would probably have put that logic into CheckInputs() itself and then have it delegate to CheckInputsParalell() or so. Are there other factors at play that I am not seeing, like keeping the changes in CheckInputs() as minimal as possible for example?

sdaftuar replied:
CheckInputs() is also invoked directly from ConnectBlock(), which handles the parallel scriptcheck threads itself.

hugohn commented on src/validation.cpp:
```diff
@@ -1672,6 +1675,23 @@ void ThreadScriptCheck(int worker_num) {
+    if (!control.Wait()) {
+        return state.Invalid(ValidationInvalidReason::CONSENSUS, false, REJECT_INVALID, "script-verify-flag-failed");
+    }
+    if (nScriptCheckThreads && !single_threaded_validation && cacheFullScriptStore) {
```
nit: since `(nScriptCheckThreads && !single_threaded_validation)` is used several times throughout this function, maybe declare it at the beginning of the function? (e.g. `bool use_multiple_threads = nScriptCheckThreads && !single_threaded_validation`).

sdaftuar replied:
Done

sdaftuar commented:
There was a silent merge conflict in the prior version of this PR ([15169.2](https://github.com/sdaftuar/bitcoin/commits/15169.2)) so I've rebased -- still intend to revisit the fix for the scriptcache and outstanding nits.

sdaftuar commented:
Updated with an improved fix to the script caching issue; @jnewbery @TheBlueMatt care to take a look?
